From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Wed, 15 Jun 2022 21:38:43 +0200
Subject: [PATCH] Add Server#setMaxPlayers


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 12553b55cf1f70d1eba06a9a0909c9d0134b4ef9..189d5e9924c02b14d0f9c9d95b6741e4f8ec745a 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -182,6 +182,18 @@ public final class Bukkit {
         return server.getMaxPlayers();
     }
 
+    // Unicorn start
+    /**
+     * Sets the maximum amount of players which can login to this server.
+     * It's only temporary and after restart this value will be reset to default value (from server.properties).
+     *
+     * @param maxPlayers the amount of players this server will allow
+     */
+    public static void setMaxPlayers(int maxPlayers) {
+        server.setMaxPlayers(maxPlayers);
+    }
+    // Unicorn end
+
     /**
      * Get the game port that the server runs on.
      *
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 175aed3917907ee800062ced58e93fd5315b3c24..b26a0c0e11225f95abed883db99e1a1a37b69474 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -153,6 +153,16 @@ public interface Server extends PluginMessageRecipient {
      */
     public int getMaxPlayers();
 
+    // Unicorn start
+    /**
+     * Sets the maximum amount of players which can login to this server.
+     * It's only temporary and after restart this value will be reset to default value (from server.properties).
+     *
+     * @param maxPlayers the amount of players this server will allow
+     */
+    public void setMaxPlayers(int maxPlayers);
+    // Unicorn end
+
     /**
      * Get the game port that the server runs on.
      *
