From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Mon, 13 Jun 2022 18:22:44 +0200
Subject: [PATCH] Skip processing move event if there is no listeners


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 8716582465d57bb8de206dd4881fcf30a6e270dd..2c8e17ee0cdb31120a4c8a9dd28d68a22189b14f 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -267,7 +267,7 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
                     this.lastPitch = to.getPitch();
 
                     // Skip the first time we do this
-                    if (true) { // Spigot - don't skip any move events
+                    if (PlayerMoveEvent.getHandlerList().getRegisteredListeners().length > 0) { // Unicorn - skip processing move event if there is no listeners // Spigot - don't skip any move events
                         Location oldTo = to.fastClone(); // Unicorn - Fast Clone
                         PlayerMoveEvent event = new PlayerMoveEvent(player, from, to);
                         this.server.getPluginManager().callEvent(event);
