From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Fri, 23 Sep 2022 21:17:20 +0200
Subject: [PATCH] Allow render distance from 1, not 3


diff --git a/src/main/java/net/minecraft/server/PlayerChunkMap.java b/src/main/java/net/minecraft/server/PlayerChunkMap.java
index 43f47ae6d47e86fc7bfd03cb5c6f09b100c844dc..f32ab4db2036537b94f70067ee6be8a8274e1034 100644
--- a/src/main/java/net/minecraft/server/PlayerChunkMap.java
+++ b/src/main/java/net/minecraft/server/PlayerChunkMap.java
@@ -275,7 +275,7 @@ public class PlayerChunkMap {
     }
 
     public void a(int i) {
-        i = MathHelper.clamp(i, 3, 32);
+        i = MathHelper.clamp(i, 1, 32); // Unicorn -> Allow distance from 1, not 3
         if (i != this.g) {
             int j = i - this.g;
             ArrayList arraylist = Lists.newArrayList(this.managedPlayers);
@@ -315,7 +315,7 @@ public class PlayerChunkMap {
 
     // PaperSpigot start - Player view distance API
     public void updateViewDistance(EntityPlayer player, int viewDistance) {
-        viewDistance = MathHelper.clamp(viewDistance, 3, 32);
+        viewDistance = MathHelper.clamp(viewDistance, 1, 32); // Unicorn -> Allow distance from 1, not 3
         if (viewDistance != player.viewDistance) {
             int cx = (int) player.locX >> 4;
             int cz = (int) player.locZ >> 4;
