From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 26 Sep 2016 01:51:30 -0400
Subject: [PATCH] Implement [Paper-0161] Disable Vanilla Chunk GC by Aikar

Bukkit has its own system for this.

diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index d5526a92d9014002a31c29ad60f0d748110726ac..887aae3c405e552599b4277c6861b4ce22b06e7f 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -925,7 +925,8 @@ public class WorldServer extends World implements IAsyncTaskHandler {
 
             this.chunkProvider.saveChunks(flag, iprogressupdate);
             // CraftBukkit - ArrayList -> Collection
-            Collection arraylist = this.chunkProviderServer.a();
+            // Paper start
+            /* Collection arraylist = this.chunkProviderServer.a();
             Iterator iterator = arraylist.iterator();
 
             while (iterator.hasNext()) {
@@ -934,8 +935,8 @@ public class WorldServer extends World implements IAsyncTaskHandler {
                 if (chunk != null && !this.manager.a(chunk.locX, chunk.locZ)) {
                     this.chunkProviderServer.queueUnload(chunk.locX, chunk.locZ);
                 }
-            }
-
+            } */
+            // Paper end
         }
     }
 
