From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Brokkonaut <hannos17@gmx.de>
Date: Tue, 7 Feb 2023 23:01:51 +0100
Subject: [PATCH] Implement [Paper-0153] Fix AIOOBE in inventory handling by
 Brokkonaut


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index d948ae6d8a3c4bc93a6e18f438513abc0a408f76..4f6291d2fff7ffd85cbe2d763f8ac2022f653c1f 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1573,7 +1573,7 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
                 } else if (packetplayinwindowclick.f() == 3) {
                     if (packetplayinwindowclick.c() == 2) {
                         click = ClickType.MIDDLE;
-                        if (packetplayinwindowclick.b() == -999) {
+                        if (packetplayinwindowclick.b() < 0) { // Paper - GH-404
                             action = InventoryAction.NOTHING;
                         } else {
                             Slot slot = this.player.activeContainer.getSlot(packetplayinwindowclick.b());
