From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lew <skycubesv@gmail.com>
Date: Sun, 4 Dec 2022 12:55:11 +0100
Subject: [PATCH] Implement [InsanePaper-390] Heavily optimize Tuinity
 controlled flush patch


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 394b51781d5883f8beb54a142a597df784cb41c7..765bd6a81ce13248f207424fed3d375eca2ee781 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -949,7 +949,7 @@ public abstract class MinecraftServer extends org.mythicprojects.unicornspigot.t
                 }
 
                 // Tuinity start - controlled flush for entity tracker packets
-                List<NetworkManager> disabledFlushes = new ArrayList<>(worldserver.players.size());
+                java.util.Set<NetworkManager> disabledFlushes = new java.util.HashSet<>(worldserver.players.size()); // InsanePaper - List -> Set
                 for (EntityHuman player : worldserver.players) {
                     if (!(player instanceof EntityPlayer)) continue; // skip non-player entities
                     PlayerConnection connection = ((EntityPlayer) player).playerConnection;
