From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Thu, 22 Jun 2023 13:39:44 +0200
Subject: [PATCH] Option to disable statistics and achievements


diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index 1ed5199837d172cf9bc9b607c92c48fd82960eff..c177ad73893223864188ef19ae8f96397c539d62 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -13,6 +13,7 @@ import java.util.List;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import org.mythicprojects.unicornspigot.config.UnicornConfig;
 // CraftBukkit start
 import org.bukkit.Bukkit;
 import org.bukkit.WeatherType;
@@ -368,6 +369,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
     }
 
     protected void i_() {
+        if (UnicornConfig.get().settings.disableStatisticsAndAchievements) return; // Unicorn
         BiomeBase biomebase = this.world.getBiome(new BlockPosition(MathHelper.floor(this.locX), 0, MathHelper.floor(this.locZ)));
         String s = biomebase.ah;
         AchievementSet achievementset = (AchievementSet) this.getStatisticManager().b((Statistic) AchievementList.L);
diff --git a/src/main/java/net/minecraft/server/StatisticManager.java b/src/main/java/net/minecraft/server/StatisticManager.java
index 5d6c3deac0d16d522ac3ef793a3fea3d74f05e68..c83eb79af8a29165ba933f078243dcfa3238cf08 100644
--- a/src/main/java/net/minecraft/server/StatisticManager.java
+++ b/src/main/java/net/minecraft/server/StatisticManager.java
@@ -2,6 +2,7 @@ package net.minecraft.server;
 
 import com.google.common.collect.Maps;
 import java.util.Map;
+import org.mythicprojects.unicornspigot.config.UnicornConfig;
 
 public class StatisticManager {
 
@@ -18,6 +19,7 @@ public class StatisticManager {
     }
 
     public void b(EntityHuman entityhuman, Statistic statistic, int i) {
+        if (UnicornConfig.get().settings.disableStatisticsAndAchievements) return; // Unicorn
         if (!statistic.d() || this.b((Achievement) statistic)) {
             // CraftBukkit start - fire Statistic events
             org.bukkit.event.Cancellable cancellable = org.bukkit.craftbukkit.event.CraftEventFactory.handleStatisticsIncrease(entityhuman, statistic, this.getStatisticValue(statistic), i);
@@ -30,6 +32,7 @@ public class StatisticManager {
     }
 
     public void setStatistic(EntityHuman entityhuman, Statistic statistic, int i) {
+        if (UnicornConfig.get().settings.disableStatisticsAndAchievements) return; // Unicorn
         StatisticWrapper statisticwrapper = (StatisticWrapper) this.a.get(statistic);
 
         if (statisticwrapper == null) {
@@ -41,6 +44,7 @@ public class StatisticManager {
     }
 
     public int getStatisticValue(Statistic statistic) {
+        if (UnicornConfig.get().settings.disableStatisticsAndAchievements) return 0; // Unicorn
         StatisticWrapper statisticwrapper = (StatisticWrapper) this.a.get(statistic);
 
         return statisticwrapper == null ? 0 : statisticwrapper.a();
diff --git a/src/main/java/org/mythicprojects/unicornspigot/config/UnicornConfig.java b/src/main/java/org/mythicprojects/unicornspigot/config/UnicornConfig.java
index 1862e0b860ad91b8585d462e6f9681343bbded08..131b584d7292027ad8d4baef68241c38a41019b4 100644
--- a/src/main/java/org/mythicprojects/unicornspigot/config/UnicornConfig.java
+++ b/src/main/java/org/mythicprojects/unicornspigot/config/UnicornConfig.java
@@ -39,6 +39,10 @@ public final class UnicornConfig extends EngineConfig<UnicornWorldConfig> {
         @Comment("Whether checking if everyone is sleeping should ignore players in end/nether.")
         public boolean sleepingIgnoreOtherDimensions = false;
 
+        @Comment("Whether to disable statistics (ESC -> Statistics) and achievements.")
+        @Comment("Enabling this option could break some plugins that rely on this mechanics.")
+        public boolean disableStatisticsAndAchievements = false;
+
     }
 
     @Comment("Optimizations that could improve performance, but may cause some issues.")
