From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Thu, 22 Jun 2023 14:12:39 +0200
Subject: [PATCH] Option to optimize armor stands


diff --git a/src/main/java/net/minecraft/server/EntityArmorStand.java b/src/main/java/net/minecraft/server/EntityArmorStand.java
index 2ea4a5fd32629a7931189786f86dce99d1db6d63..75ca5e8c78a7c998ee5d2ae85219753f1b305f48 100644
--- a/src/main/java/net/minecraft/server/EntityArmorStand.java
+++ b/src/main/java/net/minecraft/server/EntityArmorStand.java
@@ -51,7 +51,7 @@ public class EntityArmorStand extends EntityLiving {
     }
 
     public boolean bM() {
-        return super.bM() && !this.hasGravity();
+        return super.bM() && (this.world.unicornConfig.optimizations.optimizeArmorStands || !this.hasGravity()); // Unicorn
     }
 
     protected void h() {
@@ -245,6 +245,7 @@ public class EntityArmorStand extends EntityLiving {
     protected void s(Entity entity) {}
 
     protected void bL() {
+        if (this.world.unicornConfig.optimizations.optimizeArmorStands) return; // Unicorn
         List list = this.world.getEntities(this, this.getBoundingBox());
 
         if (list != null && !list.isEmpty()) {
@@ -489,6 +490,7 @@ public class EntityArmorStand extends EntityLiving {
     }
 
     public void g(float f, float f1) {
+        if (this.world.unicornConfig.optimizations.optimizeArmorStands) return; // Unicorn
         if (!this.hasGravity()) {
             super.g(f, f1);
         }
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 1dfec834a18ebb749c42a35a6323808accd3880c..2cc95f6795fefc73c9bfebdb75f4ee3894f67ac8 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1456,6 +1456,7 @@ public abstract class World implements IBlockAccess {
         // Spigot end
 
         if (!this.unicornConfig.optimizations.entitiesCollisions && (entity instanceof EntityHuman || entity instanceof EntityInsentient || entity instanceof EntityArmorStand)) return arraylist; // Unicorn
+        if (this.unicornConfig.optimizations.optimizeArmorStands && entity instanceof EntityArmorStand) return arraylist; // Unicorn
         if (entity instanceof EntityItem) return arraylist; // PaperSpigot - Optimize item movement
 
         double d0 = 0.25D;
diff --git a/src/main/java/org/mythicprojects/unicornspigot/config/UnicornWorldConfig.java b/src/main/java/org/mythicprojects/unicornspigot/config/UnicornWorldConfig.java
index 4c973b59d4a3592e068f290d9d4c3451345d4830..37f0f214de27ce74170729a771a0cda94585cf94 100644
--- a/src/main/java/org/mythicprojects/unicornspigot/config/UnicornWorldConfig.java
+++ b/src/main/java/org/mythicprojects/unicornspigot/config/UnicornWorldConfig.java
@@ -26,6 +26,9 @@ public final class UnicornWorldConfig extends EngineConfig.WorldConfig {
         @Comment("Disabling this option may improve performance a bit, but it will break entities collisions.")
         public boolean entitiesCollisions = true;
 
+        @Comment("Whether to optimize armor stands by disabling it's collisions and gravity.")
+        public boolean optimizeArmorStands = false;
+
         @Comment("Configuration of ticking specific mechanics.")
         public Ticking ticking = new Ticking();
 
