From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sat, 24 Jun 2023 04:45:28 +0200
Subject: [PATCH] Use CerberusLoader to load dependencies


diff --git a/build.gradle.kts b/build.gradle.kts
index 65bc6fb9fba7d7ed213b034ffa2327dfd4b75698..23ad78fbd87e122b5b9826ee4c4b16916307e261 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -1,3 +1,4 @@
+
 val minecraftVersion = "1_8_R3"
 
 dependencies {
@@ -5,29 +6,40 @@ dependencies {
 
     // Native (Minecraft) libraries start
     compileOnly("io.netty:netty-all:4.0.23.Final") // Loaded by CerberusLoader
-    implementation("net.sf.trove4j:trove4j:3.0.3")
     // Unicorn start - updated log4j to 2.19.0
-    implementation("org.apache.logging.log4j:log4j-api:2.19.0")
-    implementation("org.apache.logging.log4j:log4j-core:2.19.0")
+    library("org.apache.logging.log4j:log4j-api:2.19.0")
+    library("org.apache.logging.log4j:log4j-core:2.19.0")
     // Unicorn end - updated log4j to 2.19.0
-    implementation("jline:jline:2.12.1")
-    implementation("commons-io:commons-io:2.4")
-    implementation("commons-codec:commons-codec:1.9")
-    implementation("net.sf.jopt-simple:jopt-simple:5.0") // Unicorn - updated jopt-simple to 5.0
-    implementation("com.mojang:authlib:1.5.21")
-    implementation("org.xerial:sqlite-jdbc:3.7.2")
+    library("commons-io:commons-io:2.4")
+    library("commons-codec:commons-codec:1.9")
+    library("net.sf.jopt-simple:jopt-simple:5.0") // Unicorn - updated jopt-simple to 5.0
+    implementation("com.mojang:authlib:1.5.21") {
+        // Unicorn start - use our versions of dependencies
+        exclude("org.apache.commons")
+        exclude("commons-io")
+        exclude("commons-codec")
+        exclude("org.apache.logging.log4j")
+        exclude("com.google.code.gson")
+        exclude("com.google.guava")
+        // Unicorn end
+    }
+    library("org.xerial:sqlite-jdbc:3.7.2")
     compileOnly("mysql:mysql-connector-java:5.1.14") // Loaded by CerberusLoader
     // Native libraries end
 
     // UnicornSpigot libraries start
     // Paper start - Use TerminalConsoleAppender
-    implementation("net.minecrell:terminalconsoleappender:1.3.0")
-    implementation("org.jline:jline-terminal-jansi:3.21.0")
+    library("net.minecrell:terminalconsoleappender:1.3.0")
+    val jLineVersion = "3.21.0"
+    library("org.jline:jline-reader:$jLineVersion")
+    library("org.jline:jline-terminal:$jLineVersion")
+    library("org.jline:jline-terminal-jansi:$jLineVersion")
+    library("org.fusesource.jansi:jansi:2.4.0")
     // Paper end
 
     val byteBuddyVersion = "1.12.12"
-    implementation("net.bytebuddy:byte-buddy-agent:$byteBuddyVersion")
-    implementation("net.bytebuddy:byte-buddy:$byteBuddyVersion")
+    library("net.bytebuddy:byte-buddy-agent:$byteBuddyVersion")
+    library("net.bytebuddy:byte-buddy:$byteBuddyVersion")
 
     val okaeriConfigsVersion = "5.0.0-beta.6"
     implementation("eu.okaeri:okaeri-configs-core:$okaeriConfigsVersion")
@@ -36,9 +48,9 @@ dependencies {
     implementation("eu.okaeri:okaeri-configs-validator-okaeri:$okaeriConfigsVersion")
 
     implementation("com.velocitypowered:velocity-native:1.1.9")
-    implementation("com.github.ben-manes.caffeine:caffeine:2.9.3")
-    implementation("it.unimi.dsi:fastutil:8.5.11")
-    implementation("com.eatthepath:fast-uuid:0.2.0")
+    library("com.github.ben-manes.caffeine:caffeine:2.9.3")
+    library("it.unimi.dsi:fastutil:8.5.11")
+    library("com.eatthepath:fast-uuid:0.2.0")
     // UnicornSpigot libraries end
 
     testImplementation("junit:junit:4.13.2") // Unicorn - update JUnit to 4.13.2
@@ -51,16 +63,8 @@ tasks {
         transform(com.github.jengelman.gradle.plugins.shadow.transformers.Log4j2PluginsCacheFileTransformer()) // Paper - Use TerminalConsoleAppender
         archiveClassifier.set("") // without "-all"
 
-        // exclude org.fusesource.jansi:jansi
-        exclude("org/fusesource/jansi/**") // Loaded by CerberusLoader
-
-        val cbLibsPkg = "org.bukkit.craftbukkit.libs"
-
-        relocate("joptsimple", "$cbLibsPkg.joptsimple")
-        relocate("jline", "$cbLibsPkg.jline")
-        relocate("org.ibex", "$cbLibsPkg.org.ibex")
-        relocate("org.gjt", "$cbLibsPkg.org.gjt")
-        relocate("eu.okaeri", "$cbLibsPkg.eu.okaeri")
+        val libsPackage = "org.mythicprojects.unicornspigot.libs"
+        relocate("eu.okaeri", "$libsPackage.eu.okaeri")
 
         relocate("org.bukkit.craftbukkit", "org.bukkit.craftbukkit.v${minecraftVersion}") {
             exclude("org.bukkit.craftbukkit.Main*") // don't relocate main class
@@ -81,15 +85,15 @@ tasks {
         manifest {
             val gitHash = project.parent!!.ext["gitHash"]
             attributes(
-                "Main-Class" to "org.bukkit.craftbukkit.Main",
-                "Implementation-Title" to "CraftBukkit",
-                "Implementation-Version" to "git-PaperSpigot-$gitHash",
-                "Implementation-Vendor" to "Bukkit Team",
-                "Specification-Title" to "Bukkit",
-                "Specification-Version" to project.version,
-                "Specification-Vendor" to "Bukkit Team",
-                "Multi-Release" to "true"
+                    "Main-Class" to "org.bukkit.craftbukkit.Main",
+                    "Implementation-Title" to "CraftBukkit",
+                    "Implementation-Version" to "git-PaperSpigot-$gitHash",
+                    "Implementation-Vendor" to "Bukkit Team",
+                    "Specification-Title" to "Bukkit",
+                    "Specification-Version" to project.version,
+                    "Specification-Vendor" to "Bukkit Team",
+                    "Multi-Release" to "true"
             )
         }
     }
-}
+}
\ No newline at end of file
diff --git a/src/main/java/org/mythicprojects/unicornspigot/util/Objects.java b/src/main/java/org/mythicprojects/unicornspigot/util/Objects.java
index c62f27450537555e0ae72035d4b770d8c3653373..0362c0309c53767f62461fabffef8cdcc20442c2 100644
--- a/src/main/java/org/mythicprojects/unicornspigot/util/Objects.java
+++ b/src/main/java/org/mythicprojects/unicornspigot/util/Objects.java
@@ -101,7 +101,7 @@ public final class Objects {
             return this.addHolder(name, String.valueOf(value));
         }
 
-        public ToStringHelper addValue(@javax.annotation.Nullable Object value) {
+        public ToStringHelper addValue(@Nullable Object value) {
             return this.addHolder(value);
         }
 
