From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sat, 24 Jun 2023 17:01:12 +0200
Subject: [PATCH] Make chunks I/O threads count configurable


diff --git a/src/main/java/org/bukkit/craftbukkit/chunkio/ChunkIOExecutor.java b/src/main/java/org/bukkit/craftbukkit/chunkio/ChunkIOExecutor.java
index 69e5da4c65da46ad9d1456266cd4ad15b136f6e4..183e12f04019e1907a6d9ea8ca846d8d07440fe6 100644
--- a/src/main/java/org/bukkit/craftbukkit/chunkio/ChunkIOExecutor.java
+++ b/src/main/java/org/bukkit/craftbukkit/chunkio/ChunkIOExecutor.java
@@ -5,10 +5,13 @@ import net.minecraft.server.ChunkProviderServer;
 import net.minecraft.server.ChunkRegionLoader;
 import net.minecraft.server.World;
 import org.bukkit.craftbukkit.util.AsynchronousExecutor;
+import org.mythicprojects.unicornspigot.config.UnicornConfig;
 
 public class ChunkIOExecutor {
-    static final int BASE_THREADS = 2; // PaperSpigot - Bumped value
-    static final int PLAYERS_PER_THREAD = 50;
+    // Unicorn start - config
+    static final int BASE_THREADS = UnicornConfig.get().chunks.io.baseThreads;
+    static final int PLAYERS_PER_THREAD = UnicornConfig.get().chunks.io.playersPerThread;
+    // Unicorn end
 
     private static final AsynchronousExecutor<QueuedChunk, Chunk, Runnable, RuntimeException> instance = new AsynchronousExecutor<QueuedChunk, Chunk, Runnable, RuntimeException>(new ChunkIOProvider(), BASE_THREADS);
 
diff --git a/src/main/java/org/mythicprojects/unicornspigot/config/UnicornConfig.java b/src/main/java/org/mythicprojects/unicornspigot/config/UnicornConfig.java
index f0bf074df9d1a3025521aa44bfec4caaba47c982..cf83103cf4d3520350926149dbbe56b5f65c0b7d 100644
--- a/src/main/java/org/mythicprojects/unicornspigot/config/UnicornConfig.java
+++ b/src/main/java/org/mythicprojects/unicornspigot/config/UnicornConfig.java
@@ -100,6 +100,29 @@ public final class UnicornConfig extends EngineConfig<UnicornWorldConfig> {
 
     }
 
+    @Comment("Configuration of systems related to chunks.")
+    public Chunks chunks = new Chunks();
+
+    public static class Chunks extends UnicornConfigSection {
+
+        @Comment("Configuration of chunks I/O loadings.")
+        @Comment("If you're using CPU with more threads, you can try to increase this value.")
+        public IO io = new IO();
+
+        public static class IO extends UnicornConfigSection {
+
+            @Comment("The amount of threads that will be used to load chunks.")
+            public int baseThreads = 2;
+
+            @Comment("This values represents how many players are needed to create a new thread for chunks I/O.")
+            @Comment("For example, if you set this value to 25, baseThreads it's 2 and there are 75 players online,")
+            @Comment("then 3 threads will be used to load chunks.")
+            public int playersPerThread = 50;
+
+        }
+
+    }
+
     @Comment("Per world settings.")
     public Map<String, UnicornWorldConfig> worldSettings = new HashMap<>();
 
