From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sat, 24 Jun 2023 19:43:31 +0200
Subject: [PATCH] Use Paper's CachedSizeConcurrentLinkedQueue with constant
 size()


diff --git a/src/main/java/com/destroystokyo/paper/utils/CachedSizeConcurrentLinkedQueue.java b/src/main/java/com/destroystokyo/paper/utils/CachedSizeConcurrentLinkedQueue.java
new file mode 100644
index 0000000000000000000000000000000000000000..54bd172a17bc3646ace433c645308845d08d114d
--- /dev/null
+++ b/src/main/java/com/destroystokyo/paper/utils/CachedSizeConcurrentLinkedQueue.java
@@ -0,0 +1,34 @@
+package com.destroystokyo.paper.utils;
+
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.LongAdder;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class CachedSizeConcurrentLinkedQueue<E> extends ConcurrentLinkedQueue<E> {
+    private final LongAdder cachedSize = new LongAdder();
+
+    @Override
+    public boolean add(@NotNull E e) {
+        boolean result = super.add(e);
+        if (result) {
+            this.cachedSize.increment();
+        }
+        return result;
+    }
+
+    @Nullable
+    @Override
+    public E poll() {
+        E result = super.poll();
+        if (result != null) {
+            this.cachedSize.decrement();
+        }
+        return result;
+    }
+
+    @Override
+    public int size() {
+        return this.cachedSize.intValue();
+    }
+}
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index c5f4a11f1bf8ca9fa222462cf90f4f7348b0280d..152271518414a1ae81223ae27ad99479478bd48e 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -100,7 +100,7 @@ public abstract class MinecraftServer extends org.mythicprojects.unicornspigot.t
     private long X = 0L;
     private final GameProfileRepository Y;
     private final UserCache Z;
-    protected final Queue<FutureTask<?>> j = new java.util.concurrent.ConcurrentLinkedQueue<FutureTask<?>>(); // Spigot, PAIL: Rename
+    protected final Queue<FutureTask<?>> j = new com.destroystokyo.paper.utils.CachedSizeConcurrentLinkedQueue<>(); // Spigot, PAIL: Rename // Paper - Make size() constant-time
     private Thread serverThread;
     private long ab = az();
 
