From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sun, 25 Dec 2022 17:30:48 +0100
Subject: [PATCH] Backport ServerShutdownEvent


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index a43cc294ff8037d614ac69ef284ac397a7e131aa..c0b9935392d44a348b38db005aaf4ae1484996a8 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -478,6 +478,7 @@ public abstract class MinecraftServer extends org.mythicprojects.unicornspigot.t
         // CraftBukkit start - prevent double stopping on multiple threads
         synchronized(stopLock) {
             if (hasStopped) return;
+            new org.bukkit.event.server.ServerShutdownEvent(server).callEvent(); // Unicorn
             hasStopped = true;
         }
         // CraftBukkit end
