From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Sun, 25 Jun 2023 15:46:05 +0200
Subject: [PATCH] Backport modern messages sending API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 4944e18154c69cc1eaafc1e27938ff307debdf9b..43e24eb7a1840a8c506b89f052dd2b1604e3ad0e 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -20,6 +20,7 @@ import java.util.Set;
 import java.util.UUID;
 import java.util.logging.Level;
 import java.util.logging.Logger;
+import net.md_5.bungee.api.ChatMessageType;
 import net.md_5.bungee.api.chat.BaseComponent;
 
 import net.minecraft.server.*;
@@ -166,20 +167,6 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     }
 
     // Paper start
-    @Override
-    public void sendMessage(BaseComponent component) {
-        sendMessage(new BaseComponent[]{component});
-    }
-
-    @Override
-    public void sendMessage(BaseComponent... components) {
-        if (getHandle().playerConnection == null) return;
-
-        PacketPlayOutChat packet = new PacketPlayOutChat();
-        packet.components = components;
-        getHandle().playerConnection.sendPacket(packet);
-    }
-
     @Override
     public void setPlayerListHeaderFooter(BaseComponent[] header, BaseComponent[] footer) {
         PacketPlayOutPlayerListHeaderFooter packet = new PacketPlayOutPlayerListHeaderFooter();
@@ -1617,19 +1604,26 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
             return java.util.Collections.unmodifiableSet( ret );
         }
 
+        // Unicorn start
         @Override
-        public void sendMessage(BaseComponent component) {
-          sendMessage( new BaseComponent[] { component } );
+        public void sendMessage(ChatMessageType position, BaseComponent component) {
+            this.sendMessage( position, new BaseComponent[] { component } );
         }
 
         @Override
-        public void sendMessage(BaseComponent... components) {
-           if ( getHandle().playerConnection == null ) return;
-
-            PacketPlayOutChat packet = new PacketPlayOutChat();
+        public void sendMessage(ChatMessageType position, BaseComponent... components) {
+            PlayerConnection connection = getHandle().playerConnection;
+            if (connection == null ) return;
+
+            PacketPlayOutChat packet = new PacketPlayOutChat(null, (byte) position.ordinal());
+            // Action bar doesn't render colors, replace colors with legacy section symbols
+            if (position == ChatMessageType.ACTION_BAR) {
+                components = new BaseComponent[]{new net.md_5.bungee.api.chat.TextComponent(BaseComponent.toLegacyText(components))};
+            }
             packet.components = components;
-            getHandle().playerConnection.sendPacket(packet);
+            connection.sendPacket(packet);
         }
+        // Unicorn end
 
         // PaperSpigot start - Implement affects spawning API
         @Override
