From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peridot <peridot491@protonmail.com>
Date: Wed, 15 Jun 2022 21:38:24 +0200
Subject: [PATCH] Add Server#setMaxPlayers


diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index df357cc723e0d1d5264f059e0b6883f178e84dc7..43f7817ceb4e4ca9cd4e3264ab1ea83c2cbe4805 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -61,7 +61,7 @@ public abstract class PlayerList {
     private final Map<UUID, ServerStatisticManager> o;
     public IPlayerFileData playerFileData;
     private boolean hasWhitelist;
-    protected int maxPlayers;
+    public int maxPlayers; // Unicorn - make public
     private int r;
     private WorldSettings.EnumGamemode s;
     private boolean t;
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 0360398793615f86880a021c89c708b3713fb70b..1708796d14425daadd7ed0f21d893b4d7016096a 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -597,6 +597,11 @@ public final class CraftServer implements Server {
         return playerList.getMaxPlayers();
     }
 
+    @Override
+    public void setMaxPlayers(int maxPlayers) {
+        this.playerList.maxPlayers = maxPlayers;
+    }
+
     // NOTE: These are dependent on the corrisponding call in MinecraftServer
     // so if that changes this will need to as well
     @Override
